node/primitives/src/approval.rs

```diff
@@ -31,24 +31,136 @@ use sp_consensus_babe as babe_primitives;
 /// Earlier tranches of validators check first, with later tranches serving as backup.
 pub type DelayTranche = u32;
 
-/// A static context used to compute the Relay VRF story based on the
-/// VRF output included in the header-chain.
-pub const RELAY_VRF_STORY_CONTEXT: &[u8] = b"A&V RC-VRF";
+/// Static contexts use to generate randomness for v1 assignments.
+pub mod v1 {
+	/// A static context used to compute the Relay VRF story based on the
+	/// VRF output included in the header-chain.
+	pub const RELAY_VRF_STORY_CONTEXT: &[u8] = b"A&V RC-VRF";
 
-/// A static context used for all relay-vrf-modulo VRFs.
-pub const RELAY_VRF_MODULO_CONTEXT: &[u8] = b"A&V MOD";
+	/// A static context used for all relay-vrf-modulo VRFs.
+	pub const RELAY_VRF_MODULO_CONTEXT: &[u8] = b"A&V MOD";
 
-/// A static context used for all relay-vrf-modulo VRFs.
-pub const RELAY_VRF_DELAY_CONTEXT: &[u8] = b"A&V DELAY";
+	/// A static context used for all relay-vrf-modulo VRFs.
+	pub const RELAY_VRF_DELAY_CONTEXT: &[u8] = b"A&V DELAY";
 
-/// A static context used for transcripts indicating assigned availability core.
-pub const ASSIGNED_CORE_CONTEXT: &[u8] = b"A&V ASSIGNED";
+	/// A static context used for transcripts indicating assigned availability core.
+	pub const ASSIGNED_CORE_CONTEXT: &[u8] = b"A&V ASSIGNED";
 
-/// A static context associated with producing randomness for a core.
-pub const CORE_RANDOMNESS_CONTEXT: &[u8] = b"A&V CORE";
+	/// A static context associated with producing randomness for a core.
+	pub const CORE_RANDOMNESS_CONTEXT: &[u8] = b"A&V CORE";
 
-/// A static context associated with producing randomness for a tranche.
-pub const TRANCHE_RANDOMNESS_CONTEXT: &[u8] = b"A&V TRANCHE";
+	/// A static context associated with producing randomness for a tranche.
+	pub const TRANCHE_RANDOMNESS_CONTEXT: &[u8] = b"A&V TRANCHE";
+}
+
+/// A list of primitives introduced by v2.
+pub mod v2 {
+	use parity_scale_codec::{Decode, Encode};
+	use std::ops::BitOr;
+
+	use super::{CandidateIndex, CoreIndex};
+	use bitvec::{prelude::Lsb0, vec::BitVec};
+
+	/// A static context associated with producing randomness for a core.
+	pub const CORE_RANDOMNESS_CONTEXT: &[u8] = b"A&V CORE v2";
+	/// A static context associated with producing randomness for v2 multi-core assignments.
+	pub const ASSIGNED_CORE_CONTEXT: &[u8] = b"A&V ASSIGNED v2";
+	/// A static context used for all relay-vrf-modulo VRFs for v2 multi-core assignments.
+	pub const RELAY_VRF_MODULO_CONTEXT: &[u8] = b"A&V MOD v2";
+	/// A read-only bitvec wrapp to efficiently store candidate or core assignments. Each 1 bit identifies
+	/// a candidate(or core) by the bitfield bit index.
+	#[derive(Clone, Debug, Encode, Decode, Hash, PartialEq, Eq)]
+	pub struct AssignmentBitfield(BitVec<u8, bitvec::order::Lsb0>);
```

Comment:
I'd name it either `CandidateBitfield` or `CoreBitfield` to indicate that the bit are addressed by core index.
Typo: wrapper?

Reply:
typo indeed.

It would look strange to use CandidateBitfield for cores. To me it looks good enough as an assignment refers to both of them.

Reply:
I'll try to improve a this a bit